File: Problems/No.807_Max_Increase_to_Keep_City_Skyline/max_increase_to_keep_city_skyline.py
class Solution(object):
    def maxIncreaseKeepingSkyline(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """

        # parameters
        increased = 0
        max_col, max_row = [0]*len(grid[0]), [0]*len(grid)

        # start computing
        # find the maximum of each column
        for col in range(len(grid[0])):
            for row in range(len(grid)):
                max_col[col] = max(max_col[col], grid[row][col])
        
        # find the maximum of each row & start increasing
        for row in range(len(grid)):
            max_row[row] = max(grid[row])
            for col in range(len(grid[0])):
                increased += (min(max_row[row], max_col[col]) - grid[row][col])
        
        return increased

File: Problems/No.807_Max_Increase_to_Keep_City_Skyline/max_increase_to_keep_city_skyline.cpp
class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        
        // parameters
        int increased = 0;
        std::vector<int> max_col(grid[0].size(), 0), max_row(grid.size(), 0);

        // start computing
        // find the maximum of each column
        for (int col=0; col<grid[0].size(); col++)
            for (int row=0; row<grid.size(); row++)
                max_col[col] = std::max(max_col[col], grid[row][col]);

        // find the maximum of each row & start increasing
        for (int row=0; row<grid.size(); row++) {

            max_row[row] = *std::max_element(grid[row].begin(), grid[row].end());
            for (int col=0; col<grid[0].size(); col++) 
                increased += (std::min(max_row[row], max_col[col]) - grid[row][col]);
        
        }

        return increased;

    }
};
